src/Equation.jl

```diff
@@ -178,6 +192,48 @@ function copy_node(tree::Node{T})::Node{T} where {T}
     end
 end
 
+"""
+    copy_node_with_topology(
+        tree::Node{T}, id_map::IdDict{Node{T},Node{T}}
+    )::Node{T} where {T}
+
+id_map is a map from `objectid(tree)` to `copy(tree)`.
+We check against the map before making a new copy; otherwise
+we can simply reference the existing copy.
+[Thanks to Ted Hopp](https://stackoverflow.com/questions/49285475/how-to-copy-a-full-non-binary-tree-including-loops)
+"""
+function copy_node_with_topology(
+    tree::Node{T}, id_map::IdDict{Node{T},Node{T}}
+)::Node{T} where {T}
+
+    # This preserves nodes which have multiple parents.
+    # Otherwise, we would have the branch duplicated.
+    if haskey(id_map, tree)
+        return id_map[tree]
+    end
+
+    if tree.degree == 0
+        if tree.constant
+            copied_tree = Node(; val=copy(tree.val))
+        else
+            copied_tree = Node(T; feature=copy(tree.feature))
+        end
+        id_map[tree] = copied_tree
+        return copied_tree
+    elseif tree.degree == 1
+        copied_tree = Node(1, false, zero(T), 0, copy(tree.op))
+        id_map[tree] = copied_tree
+        copied_tree.l = copy_node_with_topology(tree.l, id_map)
+        return copied_tree
+    else
+        copied_tree = Node(2, false, zero(T), 0, copy(tree.op))
+        id_map[tree] = copied_tree
+        copied_tree.l = copy_node_with_topology(tree.l, id_map)
```

Comment:
I actually think this part can be done *before* putting the parent node `copied_tree` into `id_map`. The only reason it would be needed afterwards is if there is a graph in the tree - but that would be invalid anyways!

Reply:
Done - now the syntax is also much cleaner since it wraps everything in a single `get!(id_map, tree) do`.